### PR TITLE
Tekton Workload Identity

### PR DESCRIPTION
**Pull request by Richard87**
TODO:
- [x] Write tests

### CODE REVIEW

File: pkg/pipeline/prepare_pipelines.go [hunk 1]
```diff
@@ -225,22 +228,76 @@ func (ctx *pipelineContext) pipelineFileExists(pipelineFilePath string) error {
...
+
+		task.ObjectMeta.Name = taskName
 		taskMap[originalTaskName] = task
 		log.Debugf("created the task %s", task.Name)
 	}
-	return taskMap
+	return taskMap, errors.Join(errs...)
+}
+
+func validateAzureSkipContainers(task *pipelinev1.Task) {
```
Comment (nilsgstrabo) [inline pkg/pipeline/prepare_pipelines.go:265]:
```suggestion
func sanitizeAzureSkipContainersAnnotation(task *pipelinev1.Task) {
```
Is this a more descriptive name?

File: pkg/pipeline/prepare_pipelines.go [hunk 2]
```diff
@@ -225,22 +228,76 @@ func (ctx *pipelineContext) pipelineFileExists(pipelineFilePath string) error {
...
+	skip := strings.Split(task.ObjectMeta.Annotations[annotations.AzureWorkloadIdentiySkipContainers], ";")
+	var updatedSkipNames []string
+
+	for _, containerName := range skip {
```
Comment (nilsgstrabo) [inline pkg/pipeline/prepare_pipelines.go:269]: Simplify; always prepend skip items with "step-". Return error if skip item does not exist in task.steps

File: pkg/pipeline/prepare_pipelines.go [hunk 3]
```diff
@@ -225,22 +228,76 @@ func (ctx *pipelineContext) pipelineFileExists(pipelineFilePath string) error {
...
+			updatedSkipNames = append(updatedSkipNames, "step-"+containerName)
+			continue
+		}
+
+		// Task containers neither containerName or step-containerName,ignore it, but log the warning
+		log.Warnf("Trying to ignore unknown container '%s' in task '%s' for Azure Workload Identity", containerName, task.Name)
+	}
+
+	// Ignore Tekton init containers:
+	tektonInitContainers := []string{"place-scripts", "prepare"}
```
Comment (nilsgstrabo) [inline pkg/pipeline/prepare_pipelines.go:291]: These should always be included in updatedSkipNames. Init updatedSkipNames with these values

File: pkg/pipeline/radix-application_config.go [hunk 4]
```diff
@@ -52,20 +52,23 @@ func (ctx *pipelineContext) createConfigMap(configFileContent string, prepareBui
...
+			OwnerReferences: []metav1.OwnerReference{*ctx.ownerReference},
 		},
-		metav1.CreateOptions{})
+		Data: map[string]string{
+			pipelineDefaults.PipelineConfigMapContent:      configFileContent,
+			pipelineDefaults.PipelineConfigMapBuildContext: string(buildContext),
+		},
+	}
+	if ctx.ownerReference != nil {
+		cm.ObjectMeta.OwnerReferences = []metav1.OwnerReference{*ctx.ownerReference}
```
Comment (nilsgstrabo) [inline pkg/pipeline/radix-application_config.go:68]: owenrReference is already set in line 60. Can ownerReference be nil? If so, line 60 should be remove

File: pkg/utils/test/setup.go [hunk 5]
```diff
@@ -2,33 +2,39 @@ package test
...
 	kubeclient := kubeclientfake.NewSimpleClientset()
-	radixclient := radixclientfake.NewSimpleClientset()
-	return kubeclient, radixclient
+	rxclient := radixclientfake.NewSimpleClientset()
+	tknclient := tektonclientfake.NewSimpleClientset()
+	return kubeclient, rxclient, tknclient
 }
 
-func (params *TestParams) ApplyRd(radixClient radixclient.Interface) *v1.RadixDeployment {
+func (params *TestParams) ApplyRd(radixClient radixclient.Interface) (*v1.RadixDeployment, error) {
```
Comment (nilsgstrabo) [inline pkg/utils/test/setup.go:24]: This method and the TestParams struct is never used. Remove it.

**Review by nilsgstrabo — CHANGES_REQUESTED** [review]
Added some suggestions

**Review by satr — CHANGES_REQUESTED** [review]
Suggested changes

File: pkg/pipeline/prepare_pipelines.go [hunk 6]
```diff
@@ -225,22 +228,60 @@ func (ctx *pipelineContext) pipelineFileExists(pipelineFilePath string) error {
...
+	for _, step := range skipSteps {
+		containsStep := slices.ContainsFunc(task.Spec.Steps, func(s pipelinev1.Step) bool {
+			return s.Name == step
+		})
+		if !containsStep {
+			errs = append(errs, fmt.Errorf("step '%s' is not defined in task '%s': %w", step, task.Name, validation.ErrSkipStepNotFound))
+		}
+	}
+
+	SkipContainers := []string{"place-scripts", "prepare"}
```
[the remark this answers is not in the document]
Comment (Richard87) [reply pkg/pipeline/prepare_pipelines.go:278]: I don't think so, these steps doesn't exist yet as they are first created in the Pod (they are init containers)

**Review by satr — APPROVED** [review]

**Review by nilsgstrabo — APPROVED** [review]

**Pull request merged**